react/components/MultiSelect/index.js

```diff
@@ -0,0 +1,197 @@
+import PropTypes from 'prop-types'
+
+import Tag from './Tag'
+import DropdownList from './DropdownList'
+
+export default class MultiSelect extends Component {
+  constructor(props) {
+    super(props)
+    this.searchInput = React.createRef()
+
+    this.state = {
+      active: false,
+      focusedOption: 0,
+      hovering: false,
+      loadCount: 0,
+      options: [],
+      searchTerm: '',
+    }
+  }
+
+  handleFocus = () => {
+    this.setState({ active: true })
+  }
+
+  handleBlur = () => {
+    if (!this.state.hovering) {
+      this.setState({ active: false })
+    }
+  }
+
+  handleKeyPress = event => {
+    if (this.state.searchTerm !== '') {
+      if (event.key === 'Tab' || event.key === 'Enter') {
+        event.preventDefault()
+        this.selectFocused()
+      } else if (event.key === 'ArrowUp') {
+        event.preventDefault()
+        this.moveFocusUp()
+      } else if (event.key === 'ArrowDown') {
+        event.preventDefault()
+        this.moveFocusDown()
+      }
+    } else if (event.key === 'Backspace') {
+      this.unselectLast()
+    }
+  }
+
+  handleSearch = event => {
+    const searchTerm = event.target.value
+    this.setState(
+      prevState => {
+        return {
+          loadCount: prevState.loadCount + 1,
+          searchTerm,
+          focusedOption: 0,
+          options: [],
+        }
+      },
+      async () => {
+        const options = await this.props.onSearch(searchTerm)
+        this.setState(prevState => {
+          return { loadCount: prevState.loadCount - 1, options }
+        })
+      }
+    )
+  }
+
+  handleSelect = index => {
+    this.props.onChange([...this.props.selected, this.state.options[index]])
+    this.setState(
+      {
+        searchTerm: '',
+      },
+      () => {
+        this.searchInput.current.focus()
+      }
+    )
+  }
+
+  handleUnselect = index => {
+    this.props.onChange(this.props.selected.filter((opt, i) => index !== i))
+    this.searchInput.current.focus()
+  }
+
+  selectFocused = () => {
+    const index = this.state.focusedOption
+    if (this.state.options.length > 0) {
+      this.handleSelect(index)
+    }
+  }
+
+  unselectLast = () => {
+    const length = this.props.selected.length
+    if (length > 0) {
+      this.handleUnselect(length - 1)
+    }
+  }
+
+  moveFocusUp = () => {
+    const newFocus = this.state.focusedOption - 1
+    if (newFocus >= 0) {
+      this.setState({ focusedOption: newFocus })
+    }
+  }
+
+  moveFocusDown = () => {
+    const newFocus = this.state.focusedOption + 1
+    if (newFocus <= this.state.options.length - 1) {
+      this.setState({ focusedOption: newFocus })
+    }
+  }
+
+  render() {
+    const tags = this.props.selected.map((tag, index) => (
+      <Tag
+        key={index}
+        onClick={() => {
+          this.handleUnselect(index)
+        }}
+      >
+        {tag}
+      </Tag>
+    ))
+    const showDropdown = this.state.active && this.state.searchTerm !== ''
```

Comment:
`isDropdownVisible`

(sobre esses nomes, `showDropdown` por exemplo poderia ser nome de uma função, `is...` ou `has...` deixa menos ambiguo)

Reply:
Boa sugestão, obrigada pela explicação, mudei.